[Woo POS] [Cash & Receipts] Modal view to send receipts after successful payment

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->
Part of #14459
Closes: #14461 

## Description

This PR adds a modal view that will be triggered when tapping on the "Receipt" button, after a successful POS transaction. The modal contains a simple textfield where we'll enter the customer email address, so we can send an email receipt. Designs and error handling are TBD.

https://github.com/user-attachments/assets/e980008b-542f-46a2-bc17-0a1c107d2631

## Testing information
- Set feature flag `.sendReceiptsForPointOfSale`  to `true`
- Log into Povilas testing site https://lovely-aware-sturgeon.jurassic.ninja/ details: PdfdoF-5Qj-p2
- Proceed with a POS transaction.
- Tap on the textfield for focus, observe the built-in keyboard can be opened (cmd+K in simulator)
- After entering a valid email address, tap on "send", check your email and confirm the email has been received.
- The modal should be dismissed automatically upon completion (it may take 2-3 seconds due to the multiple network requests that are currently necessary). UI while the user waits will be improved in following PRs.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Prompt “send receipt” modal view when button tapped

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Send receipt from POS

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -60,7 +60,7 @@ class PointOfSaleAggregateModel: ObservableObject, PointOfSaleAggregateModelProt
     @Published private(set) var orderState: PointOfSaleOrderState = .idle
 
-    private var order: Order? = nil
+    private(set) var order: Order? = nil
```
This should remain private – views should never need the order. `OrderState` should be enough, and if it's not the associated value can be updated to share more information.

I don't think you're actually reading this anywhere outside the aggregate model though...

iamgabrielma replied:
We don't, was some leftover of my previous attempt. Gone now!

joshheald commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSSendReceiptModalView.swift:
```diff
@@ -0,0 +1,103 @@
+
+struct POSSendReceiptModalView: View {
+    @EnvironmentObject private var posModel: PointOfSaleAggregateModel
+    @Environment(\.dismiss) private var dismiss
```
Is this going to stay in a sheet? If we move to a posModal this will obviously need to have a binding to an isPresented to control presentation.

iamgabrielma replied:
I'm not sure, designs are still TBD so I went ahead with presenting a view for simplicity, encapsulated to the TotalsView, if we move ahead with the design I think we can add it as `.posModal` and trigger the presentation from `posModel.shouldSomething`, which I'm guessing should also make it more testable.

joshheald commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSSendReceiptModalView.swift:
```diff
@@ -0,0 +1,103 @@
+import SwiftUI
+
+struct POSSendReceiptModalView: View {
+    @EnvironmentObject private var posModel: PointOfSaleAggregateModel
```
Totally fine and what it's meant for, but there are still arguments where we don't really need to observe state, to just pass the bits of the posModel that we actually do need in. You could replace this with a property to hold the send receipt function.

The benefit is that it becomes usable outside POS, which seems unlikely, but it's also simpler for creating the preview.

iamgabrielma replied:
So just send the email back as a callback, and then handle both the call to send the receipt and dismiss from the TotalsView? Eg:
```

// POSSendReceiptModalView:

var completion: (String) -> ()

Button(action: {
  Task { @MainActor in
    completion(textFieldInput)
  }
}

// TotalsView

.sheet(isPresented: $isShowingSendReceiptModal) {
  POSSendReceiptModalView(inputCallback: { email in
    Task { await posModel.sendReceipt(to: email)  } 
  })
}
```

joshheald replied:
Yep, but I'd call it `sendReceipt`, not `completion`, on the modal too.

iamgabrielma replied:
Gotcha, done: https://github.com/woocommerce/woocommerce-ios/pull/14481/commits/3169cbb7a50eb4e17a8d33dd50b14f11765a7d51

iamgabrielma committed:
Merge branch 'trunk' into task/14461-pos-receipt-delivery-view

# Conflicts:
#	WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift
#	WooCommerce/Classes/POS/Utils/POSOrderPreviewService.swift
#	Yosemite/Yosemite/Tools/POS/POSOrderService.swift

iamgabrielma committed:
Make compile with updated design changes

iamgabrielma commented on WooCommerce/Classes/POS/Controllers/PointOfSaleOrderController.swift:
```diff
@@ -68,6 +69,15 @@ final class PointOfSaleOrderController: PointOfSaleOrderControllerProtocol {
+        do {
+            try await orderService.sendOrderReceipt(order: order, toEmailAddress: toEmailAddress)
+        } catch {
+            // TODO:
+            debugPrint("Error: \(error)")
```
I left this unhandled for the moment, do we want to switch the `orderState` to `.error` here? As the Order might be completed just fine, but the receipt could fail after 🤔

joshheald replied:
Probably not; we need to be really careful not to give the impression that the payent has failed, when actually only sending the receipt has.

Potentially this would be a good place for the view to handle the thrown errors somehow; then we can more easily display the error in-line with the receipt call that they relate to. This would probably need a specific view that can handle them, rather than the default button.

In the IPP version of this, the button turns into a spinner while it's sending, then a check mark when it's done. Equally, it could show the error and a retry button if it fails.

iamgabrielma replied:
That makes sense, I already have a logged issue for this case: https://github.com/woocommerce/woocommerce-ios/issues/14464. I'll see what design comes up with before making any UI changes here.

iamgabrielma committed:
localized strings

iamgabrielma committed:
Extract Color duplication into reusable func

iamgabrielma committed:
lint

iamgabrielma committed:
Make test target compile

iamgabrielma commented:
This is ready for review @joshheald , I updated the design to account for the usage of `POSOrderService` and `PointOfSaleOrderController` since my last work on this. Designs are still TBD, but it works exactly as before. Let me know if the approach makes sense, happy to iterate further 👍

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -196,6 +196,16 @@ extension PointOfSaleAggregateModel {
 
+    @MainActor
+    func sendReceipt(to emailAddress: String) async {
+        guard let order = orderController.order else {
+            return
+        }
```
Could you use orderState here? As per the deprecation note, we'll want to remove the `order` property as it's not the ideal source of truth. It may be that the view is passing the order, because it should (presumably) only be showing the receipt button when there's an order loaded, not when it's syncing or errored?

iamgabrielma replied:
The deprecation notice throw me off a bit here, as I don't see right away how we can use it to retrieve the order, so I though was still ongoing: The `orderState` property is of type `PointOfSaleOrderState`, which in its `.loaded` case only contains several strings we display in the view, and not the `order` object itself, we can get this from `PointOfSaleInternalOrderState` but I didn't wanted to use much combine since we're trying to get away from it (maybe not in this context, since it's internal observation only?)

I've made the changes here: https://github.com/woocommerce/woocommerce-ios/pull/14481/commits/e7e7a76d1523647a3595a352692f1e94439a6b57, let me know if this is what you had in mind, in which case we can change the collectPayment implementation as well on a different PR, or I can revert it and iterate.

joshheald replied:
Yeah, the deprecation notice is written in the context of the OrderController. It's saying "Hey, don't use my order property, go with my `orderState` property instead. Sorry that it caused confusion though.

I want to avoid the view having the full order, as much as possible. That does mean we have stateful controllers that manage our state, not just stateless services. If we wanted to get rid of the controllers, we would need to expose something about the order to the view, but the full model object is too much. If the view has the full `Order`, we will end up adding stuff which just accesses properties on the order and ends up wild.

If we did go the route of exposing something about the order for the view, we'd probably just have the view know about the OrderID, and then the receipt code would fetch the order before adding the email address and requesting the receipt email. E.g. the view would call `posModel.sendReceipt(for: orderID, to: emailAddress)`. For now, we have the state so let's just leave off the order/orderID, from the view's perspective.

I don't think we need to change the `collectPayment` implementation either; see my inline note, we can just rely on the order state inside the orderController, by having the orderController pass the order along to the receipt code.

joshheald commented on Yosemite/Yosemite/Tools/POS/POSOrderService.swift:
```diff
@@ -58,6 +63,17 @@ public final class POSOrderService: POSOrderServiceProtocol {
+
+    public func sendOrderReceipt(order: Order, toEmailAddress: String) async throws {
+        guard order.billingAddress?.email == nil || order.billingAddress?.email == "" else {
+            throw NSError(domain: "Email already set", code: 0)
```
It would be good to define an `Error` enum for this, rather than using NSError.

iamgabrielma replied:
True! no reason to leave it for later. Changed: https://github.com/woocommerce/woocommerce-ios/pull/14481/commits/952aef8f9d17072051292482c69e599b719bd909

joshheald commented on Yosemite/Yosemite/Tools/POS/POSOrderService.swift:
```diff
@@ -58,6 +63,17 @@ public final class POSOrderService: POSOrderServiceProtocol {
+        guard order.billingAddress?.email == nil || order.billingAddress?.email == "" else {
+            throw NSError(domain: "Email already set", code: 0)
+        }
+        let updatedBillingAddress = order.billingAddress?.copy(email: toEmailAddress)
```
The name of the `toEmailAddress` parameter makes this line difficult to read. Reading it like a sentence, it appears to say "copy the billing address to the email address", when actually it's more like the opposite of that.

iamgabrielma replied:
Changed to `recipientEmail`: https://github.com/woocommerce/woocommerce-ios/pull/14481/commits/781b3ce8cdc6a36ac33b93ef6ef47516acdf5f37

joshheald commented on Networking/Networking/Remote/ReceiptRemote.swift:
```diff
@@ -50,6 +50,21 @@ public final class ReceiptRemote: Remote {
+                    continuation.resume(throwing: error)
+                }
+            }
+        }
```
Sinc `sendReceipt` is already async, there's no need for the continuation here. However, there also doesn't seem to be any need for this function at all... will there be something different about POS receipts that needs a separate function soon? If not, let's just use `sendReceipt` directly.

iamgabrielma replied:
> Sinc sendReceipt is already async, there's no need for the continuation here. 

Oh! Updated: https://github.com/woocommerce/woocommerce-ios/pull/14481/commits/583f8fc1d11da2f93ab2aa8409183b1f59b64f04

> there also doesn't seem to be any need for this function at all.

Good question, is basically for protocol conformance: I'm not sure how we can get rid of it after the changes to the `POSOrderService` initializer to use an abstraction for `POSOrdersRemoteProtocol`. In order to follow the same pattern I'm passing a `POSReceiptsRemoteProtocol`, which forces us to have this method in the remote so we can use it through the interface later on 🤔

joshheald replied:
> In order to follow the same pattern I'm passing a POSReceiptsRemoteProtocol, which forces us to have this method in the remote so we can use it through the interface later on 🤔

I'm not sure whether you need to totally follow the protocol that way, but even if you want to, you can still name the function `sendReceipt` on the protocol, then you won't need to have an extra wrapping function in your protocol conformance on the Remote. You can just say the remote conforms to the POSReceiptsRemoteProtocol, and since it's already got `sendReceipt`, it does conform.

joshheald commented on WooCommerce/Classes/POS/Controllers/PointOfSaleOrderController.swift:
```diff
@@ -12,6 +12,7 @@ protocol PointOfSaleOrderControllerProtocol {
     var order: Order? { get }
 
     func syncOrder(for cartProducts: [CartItem], retryHandler: @escaping () async -> Void) async
+    func sendOrderReceipt(order: Order, toEmailAddress: String) async
```
This doesn't really interact with the order state, i.e. it just uses an order but doesn't play any part in mutating it or updating the source of truth. Can it be put on a different controller or be dealt with using a direct, stateless, service call?

iamgabrielma replied:
Thanks! This is the same case as https://github.com/woocommerce/woocommerce-ios/pull/14481#discussion_r1866978975, if I don't add the method to the protocol here, then we need another way to call it through the `POSAggregateModel` as we need access through the interface. My understanding is that we'd use the controller to interact with these? What would be preferable here rather than using the order controller?
- Pass the OrderService to the POSAggregateModel?
- Pass a new receiptsController protocol? (this approach seems to make more sense, as we already have items controller, and order controller)

joshheald commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSSendReceiptModalView.swift:
```diff
@@ -0,0 +1,84 @@
+                    .font(.headline)
+
+                TextField(Localization.textfieldPlaceholder, text: $textFieldInput)
+                    .keyboardType(.emailAddress)
+                    .textInputAutocapitalization(.none)
+                    .autocorrectionDisabled()
```
Well done, it's so annoying when this kind of detail is forgotten.

joshheald commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSSendReceiptModalView.swift:
```diff
@@ -0,0 +1,84 @@
+                        Text(Localization.buttonTitle)
+                            .font(Constants.buttonFont)
+                    }
+                    .frame(minWidth: UIScreen.main.bounds.width / 2)
```
Perhaps just use `posModalSizing` rather than adding another place where we rely on the screen bounds?

iamgabrielma replied:
Done! https://github.com/woocommerce/woocommerce-ios/pull/14481/commits/9547ed82bc4d8487fdf9819e99ef20afda3cf68e

joshheald commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSSendReceiptModalView.swift:
```diff
@@ -0,0 +1,84 @@
+            .padding(.top)
+            .padding()
+
+            Button(action: {
+                dismiss()
+            }) {
+                Image(systemName: "xmark")
+                    .foregroundColor(.gray)
+                    .font(.title2)
+                    .padding()
+            }
```
This, along with other stuff, could be removed by using a POSModal, and it would probably look more consistent with the rest of POS. You can still present it from the TotalsView with the same binding.

iamgabrielma replied:
True, I made the change to use `.posModal` on https://github.com/woocommerce/woocommerce-ios/pull/14481/commits/9547ed82bc4d8487fdf9819e99ef20afda3cf68e

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -196,6 +196,16 @@ extension PointOfSaleAggregateModel {
+            return
+        }
+        // TODO:
+        // Add eligiblity for correct WC and WCPay versions
```
This will need to be added before the feature flag is enabled; at the moment, on unsupported versions you tap the button and nothing happens.

joshheald reviewed (approved):
Thanks for the updates; various comments in line, approved so that you can keep making progress since this is a partial PR for feature-flagged work.

>     The modal should be dismissed automatically upon completion (it may take 2-3 seconds due to the multiple network requests that are currently necessary). UI while the user waits will be improved in following PRs.

Unfortunately this doesn't happen for me... the modal remains visible and multiple taps on `Send` result in multiple emailed receipts. It also could do with a spinner state so that the user can tell their action is being handled.

iamgabrielma committed:
remove usage of order property in favour of state

iamgabrielma committed:
make receipt view to be presented as .posModal

iamgabrielma committed:
remove posmodal usage in favor of string callback

iamgabrielma committed:
define POSOrderServiceError

iamgabrielma committed:
update parameter name for readability

iamgabrielma committed:
remove usage of continuation

Since sendReceipt(siteID:orderID:) is an async function that throws we can await its result directly in another async function.

iamgabrielma committed:
make test target compile

iamgabrielma commented:
Thanks for the thoughtful review Josh 🙇 I addressed most of your comments except a couple follow-up questions regarding the design that I'll address separately, so I can merge this for the moment.

> Unfortunately this doesn't happen for me... the modal remains visible and multiple taps on Send result in multiple emailed receipts. It also could do with a spinner state so that the user can tell their action is being handled.

I've resolved this by moving the presentation to a .posModal with a closing button, so now it dismisses when the user taps outside the area or the "x" button. I'll handle the multiple calls and UI on https://github.com/woocommerce/woocommerce-ios/issues/14480

joshheald commented on WooCommerce/Classes/POS/Presentation/TotalsView.swift:
```diff
@@ -78,8 +78,13 @@ struct TotalsView: View {
+            POSSendReceiptModalView(sendReceipt: { email in
+                Task {
+                    await posModel.sendReceipt(to: email)
+                }
```
Is this OK to run off the MainActor?

iamgabrielma replied:
Fixed: https://github.com/woocommerce/woocommerce-ios/pull/14481/commits/0c3e091148f0bef31134603b24930a24afa19a88

iamgabrielma committed:
Merge branch 'trunk' into task/14461-pos-receipt-delivery-view

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -198,12 +200,11 @@ extension PointOfSaleAggregateModel {
-        await orderController.sendOrderReceipt(order: order, toEmailAddress: emailAddress)
+        if case let .loaded(_, order) = internalOrderState {
+            await orderController.sendOrderReceipt(order: order, recipientEmail: emailAddress)
+        }
```
Since you're doing this on the `orderController`, you don't need to observe the `internalOrderState` again in the aggregate model; just make this function:

```suggestion
        await orderController.sendReceipt(recipientEmail: emailAddress)
```

And get the order within the orderController, which is fine to access its internal state.

iamgabrielma replied:
I see what you mean now, modified on 8af330cb2d37a05be4682252b30ab3bfb381e5e8

joshheald commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSSendReceiptModalView.swift:
```diff
@@ -1,10 +1,10 @@
 struct POSSendReceiptModalView: View {
-    @EnvironmentObject private var posModel: PointOfSaleAggregateModel
-    @Environment(\.dismiss) private var dismiss
+    private(set) var sendReceipt: (String) -> ()
```
This can probably be a `let`... I don't think the closure should change over time, for a single receipt modal?

iamgabrielma replied:
It can! Updated: e2a9bb3dcdb380e6ab67ab058323eb6ee46fbc3f

joshheald commented on Networking/Networking/Remote/ReceiptRemote.swift:
```diff
@@ -52,16 +52,7 @@ public final class ReceiptRemote: Remote {
 
 extension ReceiptRemote: POSReceiptsRemoteProtocol {
     public func sendPOSReceipt(siteID: Int64, orderID: Int64) async throws {
```
If you change the name of this function requirement to match the existing `sendReceipt` function you're wrapping, you can avoid having to redefine this function too...

iamgabrielma replied:
I spent more time than needed here: On one end since we use the endpoint just for POS I wanted to make it clear in the function name and wrap it, on other end, there's really no reason to. Updated on 16218bdf24db8c26b5c0db82716e7defd480a27f

joshheald reviewed (commented):
A few more notes attached – I've disabled auto merge because I'd prefer we don't merge the double-subscription to the order state in the aggregate model; it shouldn't be needed with my suggestions here

iamgabrielma committed:
Merge branch 'trunk' into task/14461-pos-receipt-delivery-view

iamgabrielma committed:
Make sendreceipt run in main actor

iamgabrielma committed:
make let

iamgabrielma committed:
remove internal state subscription from aggregate model

iamgabrielma committed:
remove unnecessary remote wrapper
